File: sample_source_code/CPP/tests/DogTest.cpp
#include <gtest/gtest.h>
#include "Dog.h"
#include "Cat.h"

// Minimal in-test visitor for verifying double-dispatch
struct CountingVisitor : public IVisitor {
    int dogVisits = 0;
    int catVisits = 0;
    void visit(Dog&) override { dogVisits++; }
    void visit(Cat&) override { catVisits++; }
};

class DogTest : public ::testing::Test {
protected:
    Dog labrador  {"Rex",   3, "Labrador",  30.0};
    Dog chihuahua {"Tiny",  1, "Chihuahua",  2.5};
    Dog defaultBreed{"Buddy", 8, "Poodle",  15.0};
};

// ── basic properties ──────────────────────────────────────────────────────────

TEST_F(DogTest, GetBreed_ReturnsCorrectBreed) {
    EXPECT_EQ("Labrador", labrador.getBreed());
}

TEST_F(DogTest, Sound_ReturnsWoof) {
    EXPECT_EQ("Woof", labrador.sound());
}

TEST_F(DogTest, GetName_ReturnsConstructorValue) {
    EXPECT_EQ("Rex", labrador.getName());
}

TEST_F(DogTest, GetAge_ReturnsConstructorValue) {
    EXPECT_EQ(3, labrador.getAge());
}

TEST_F(DogTest, GetSpecies_ReturnsCanisLupusFamiliaris) {
    EXPECT_EQ("Canis lupus familiaris", labrador.getSpecies());
}

TEST_F(DogTest, GetWeight_ReturnsConstructorValue) {
    EXPECT_DOUBLE_EQ(30.0, labrador.getWeight());
}

TEST_F(DogTest, Describe_ContainsNameAndSound) {
    const std::string desc = labrador.describe();
    EXPECT_NE(std::string::npos, desc.find("Rex"));
    EXPECT_NE(std::string::npos, desc.find("Woof"));
}

// ── computeHealthScore ────────────────────────────────────────────────────────

TEST_F(DogTest, HealthScore_IdealWeightYoungDog_NearHundred) {
    // chihuahua at ideal weight 2.5 kg, age 1 → very high score
    EXPECT_GE(chihuahua.computeHealthScore(), 90.0);
}

TEST_F(DogTest, HealthScore_IsWithinZeroToHundred) {
    EXPECT_GE(labrador.computeHealthScore(),    0.0);
    EXPECT_LE(labrador.computeHealthScore(),  100.0);
    EXPECT_GE(defaultBreed.computeHealthScore(),  0.0);
    EXPECT_LE(defaultBreed.computeHealthScore(), 100.0);
}

TEST_F(DogTest, HealthScore_OlderDog_LowerThanYounger) {
    Dog young{"A",  2, "Labrador", 30.0};
    Dog old  {"B", 12, "Labrador", 30.0};
    EXPECT_LT(old.computeHealthScore(), young.computeHealthScore());
}

TEST_F(DogTest, HealthScore_GermanShepherdIdealWeight_HighScore) {
    Dog gs{"Max", 3, "German Shepherd", 35.0};
    EXPECT_GE(gs.computeHealthScore(), 85.0);
}

// ── visitor double-dispatch ───────────────────────────────────────────────────

TEST_F(DogTest, Accept_IncrementsDogCountOnVisitor) {
    CountingVisitor v;
    labrador.accept(v);
    EXPECT_EQ(1, v.dogVisits);
    EXPECT_EQ(0, v.catVisits);
}

TEST_F(DogTest, Accept_MultipleDogs_CountsCorrectly) {
    CountingVisitor v;
    labrador.accept(v);
    chihuahua.accept(v);
    defaultBreed.accept(v);
    EXPECT_EQ(3, v.dogVisits);
    EXPECT_EQ(0, v.catVisits);
}

File: sample_source_code/typescript/tests/dog.test.ts
import { Dog } from '../src/Dog';
import type { IVisitor } from '../src/IVisitor';
import type { Cat } from '../src/Cat';

/** Minimal visitor stub for dispatch tests. */
class MockVisitor implements IVisitor {
  visitDogCalled = false;
  visitCatCalled = false;
  visitDog(_dog: Dog): void { this.visitDogCalled = true; }
  visitCat(_cat: Cat): void { this.visitCatCalled = true; }
}

describe('Dog properties', () => {
  const rex = new Dog('Rex', 3, 'Labrador', 30.0);

  test('name returns correct value', () => {
    expect(rex.name).toBe('Rex');
  });
  test('species returns Canis lupus familiaris', () => {
    expect(rex.species).toBe('Canis lupus familiaris');
  });
  test('age returns correct value', () => {
    expect(rex.age).toBe(3);
  });
  test('weight returns correct value', () => {
    expect(rex.weight).toBe(30.0);
  });
  test('breed is accessible', () => {
    expect(rex.breed).toBe('Labrador');
  });
  test('sound returns Woof', () => {
    expect(rex.sound()).toBe('Woof');
  });
  test('describe contains name and sound', () => {
    const d = rex.describe();
    expect(d).toContain('Rex');
    expect(d).toContain('Woof');
  });
  test('toString equals describe', () => {
    expect(rex.toString()).toBe(rex.describe());
  });
});

describe('Dog.computeHealthScore', () => {
  test('young dog (age <= 2) at ideal weight has score > 90', () => {
    expect(new Dog('Puppy', 1, 'Labrador', 30.0).computeHealthScore()).toBeGreaterThan(90);
  });
  test('prime dog (age <= 7) at ideal weight has score > 85', () => {
    expect(new Dog('Prime', 5, 'Labrador', 30.0).computeHealthScore()).toBeGreaterThan(85);
  });
  test('senior dog (age > 7) scores lower than prime dog', () => {
    const prime  = new Dog('Prime',  5, 'Labrador', 30.0);
    const senior = new Dog('Senior', 10, 'Labrador', 30.0);
    expect(senior.computeHealthScore()).toBeLessThan(prime.computeHealthScore());
  });
  test('overweight dog scores lower than ideal-weight dog', () => {
    const ideal = new Dog('Ideal', 3, 'Labrador', 30.0);
    const heavy = new Dog('Heavy', 3, 'Labrador', 50.0);
    expect(heavy.computeHealthScore()).toBeLessThan(ideal.computeHealthScore());
  });
  test('unknown breed uses default ideal weight', () => {
    const score = new Dog('Mixed', 3, 'Unknown Breed', 15.0).computeHealthScore();
    expect(score).toBeGreaterThanOrEqual(0);
    expect(score).toBeLessThanOrEqual(100);
  });
  test('score is always clamped to [0, 100]', () => {
    const score = new Dog('Extreme', 15, 'Labrador', 100.0).computeHealthScore();
    expect(score).toBeGreaterThanOrEqual(0);
    expect(score).toBeLessThanOrEqual(100);
  });
  test('breed matching is case-insensitive', () => {
    const lower = new Dog('Rex', 3, 'labrador', 30.0);
    const upper = new Dog('Rex', 3, 'LABRADOR', 30.0);
    expect(lower.computeHealthScore()).toBeCloseTo(upper.computeHealthScore(), 5);
  });
});

describe('Dog.accept (visitor dispatch)', () => {
  test('routes to visitDog, not visitCat', () => {
    const dog = new Dog('Buddy', 3, 'Labrador', 28.5);
    const mock = new MockVisitor();
    dog.accept(mock);
    expect(mock.visitDogCalled).toBe(true);
    expect(mock.visitCatCalled).toBe(false);
  });
});
